### ISSUE TITLE
RX 6600 prefers a lot less threads for amoebapme

### ISSUE THREAD

**Issue by bdenhollander** [openmm/openmm#3885]
Adjusting the thread block multiplier has varying results on an RDNA2 based Radeon RX 6600. Some of the benchmarks improve with more threads while others prefer less. One test that really prefers less is amoebapme.

6x multiplier: `4.23576 ns/day`
1x multiplier: `4.83736 ns/day`
Speedup: `1.14`

https://github.com/openmm/openmm/blob/4068b9151cef813d83139a8b3d877ee40941c241/platforms/opencl/src/OpenCLContext.cpp#L254-L262

CLtracer shows time spent on `gridSpreadInducedDipoles` dropping from 1st down to 6th. Would it be possible to tweak the number of threads for just that kernel to avoid regressions on other tests?

Before:
![image](https://user-images.githubusercontent.com/44237618/210470275-b987f8e4-7a26-4f09-8d49-a2f3de8d6eea.png)

After:
![image](https://user-images.githubusercontent.com/44237618/210470363-8472daa0-1a71-4357-bd1b-72f587c22efc.png)

**Comment by peastman** [comment]
Yes, it's possible to control the number of thread blocks used for a single kernel.  Here is where that kernel is executed:

https://github.com/openmm/openmm/blob/ae686364ed88dfe3d9c645db08b67700ee782281/plugins/amoeba/platforms/common/src/AmoebaCommonKernels.cpp#L1185

The argument is the maximum number of threads to use.  There also is an optional second argument for the thread block size.  If it's omitted, it uses 64 as a default size.

In this case we pass the number of atoms as the maximum number of threads.  That's to limit it on small systems, since any threads beyond the number of atoms would just be wasted.  The actual number of threads used is the minimum of the value specified there and the device-specific number chosen in the OpenCLContext constructor:

https://github.com/openmm/openmm/blob/4068b9151cef813d83139a8b3d877ee40941c241/platforms/opencl/src/OpenCLContext.cpp#L671

The challenge is to find a number that will be optimal for your current system on your current GPU.  It's interesting that it makes such a big difference in this particular case.  Can you tell why?

**Comment by bdenhollander** [comment]
> The challenge is to find a number that will be optimal for your current system on your current GPU. It's interesting that it makes such a big difference in this particular case. Can you tell why?

My guess would be the `for` loop inside `gridSpreadInducedDipoles` that is only processing 1-2 atoms per invocation since amoebapme has 23558 atoms. Inside that loop are three layers of nested `for` loops all writing to the global `pmeGrid`.

RX 6600 has 14 dual compute units containing 4x 32 wide SIMD for a total of 1792 shaders.
```
14 * 4 * 6 = 336 threads * 64 thread block size = 21504 wavefronts
14 * 4 * 1 =  56 threads * 64 thread block size =  3584 wavefronts
```

| Multiplier | Wavefronts  | Atoms/wavefront  | Loop iterations/wavefront | Wavefronts/shader |
|---|---|---|---|---|
| 6x | 21504 |  1.0955 | 1-2 | 12 |
| 1x  | 3584  | 6.5731  | 6-7 | 2 |

**Comment by peastman** [comment]
A wavefront consists of 32 threads.  To get the number of wavefronts you multiply the number of thread blocks by 2, not 64.  Every wavefront is guaranteed to be full (except the very last one, if the number of atoms isn't a multiple of 32).  Its 32 threads process 32 atoms.  Or more if they have to loop multiple times, but that only affects the outermost loop.

**Comment by bdenhollander** [comment]
Is this terminology more correct?
```
14 * 4 * 6 = 336 threads * 64 thread block size = 21504 work items
21504 / 64 group size = 336 work groups 
336 * 2 waves/work item = 672 wavefronts

14 * 4 * 1 = 56 threads * 64 thread block size = 3584 work items
3584 / 64 group size = 56 work groups 
56 * 2 waves/work item = 112 wavefronts
```

23558 atoms / 32 ~ 737 tiles

| Multiplier | Wavefronts  | Tiles/invocation |
|---|---|---|
| 6x | 672 | 1.0955 |
| 1x  | 112  | 6.5731 |

**Comment by peastman** [comment]
> 14 * 4 * 6 = 336 threads

336 thread blocks (or work groups, they're synonyms).  But the question is, why is it so much faster in one case than the other?  Does the profiler provide any information about what the bottleneck is in that kernel?

**Comment by bdenhollander** [comment]
My attempts to profile RDNA on Windows have been similar to what's reported here: https://github.com/GPUOpen-Tools/radeon_gpu_profiler/issues/56

I could get it to profile a benchmark run of gbsa but it would keep disconnecting from python.exe and disable the "Capture Profile" button for amoebapme. However, if run a full benchmark run the button stays available so I modified benchmark.py to just run gbsa and amoebapme when no test is specified.

6x
![image](https://user-images.githubusercontent.com/44237618/210699066-5501b1d1-ec93-4ef6-831a-40eb3e70e910.png)

1x
![image](https://user-images.githubusercontent.com/44237618/210699585-811117c9-9aea-45c6-ba17-22f6438acbe8.png)

| Multiplier | Wavefronts  | Average wavefront duration (µs)  | Total duration (ms) |
|---|---|---|---|
| 6x | 672 | 494.385 | 332.226 |
| 1x  | 112  | 699.004 |  78.288 |

**Comment by bdenhollander** [comment]
@peastman Kernels that use `getNumAtoms()` while having low occupancy due to high register or LDS use run faster with lower thread counts on RDNA2. These three run the most often during `amoebapme` and improve with less threads.
- `computeInducedPotentialFromGrid`
- `gridSpreadFixedMultipoles`
- `gridSpreadInducedDipoles`

I added a method to request a threadblock amount that to launch 1 threadblock per SIMD in https://github.com/openmm/openmm/commit/19b10b3ef1930c94d75565238101b4cd00e2dcdd with results below.

```
benchmark.py --platform=OpenCL --precision=single --seconds=15 --test=amoebapme
```

|   | OpenMM<br>8.0 ns/day | Tuned<br>ns/day  | Change |
|----|----|----|----|
| amoebapme | 4.20052 | 4.92702 | 17.3% |

**Comment by peastman** [comment]
The reason it passes `getNumAtoms()` is as a sanity check on the maximum number of threads.  The kernels in question have threads loop over atoms and process one at a time.  It would never make sense to have more threads than atoms, because the extra threads would have nothing to do.  So whatever else we do to optimize threads, we still need to cap it at the number of atoms.

**Comment by bdenhollander** [comment]
I rebased my [RDNA2-amoebapme](https://github.com/bdenhollander/openmm/tree/RDNA2-amoebapme) to include VkFFT and removed multiplier changes that were in the original version. 1890b595cf5ca75e75bb32108324b67dffe10daf is still showing a >10% improvement.

RDNA devices will use less threads because of the changes in OpenCLContext. Other GPUs may use less threads on large systems due to `min()` added in AmoebaCommonKernels.cpp. OpenCL benchmarks from other GPUs confirming whether this is beneficial would be appreciated.

|                                    | Original<br />ns/day | Tuned<br />ns/day | Change |
|------------------------------------|----------|---------|--------|
| amoebapme<br />CPU idle            | 3.91907  | 4.42293 | 12.86% |
| amoebapme<br />CPU background load | 4.661    | 5.50223 | 18.05% |
